### PR TITLE
Set default dataDirHostPath to /var/lib/rook

### LINKED ISSUE #1155: Set a default dataDirHostPath
For mon and osd pods to support restart, the `dataDirHostPath` must currently be set on the cluster.yaml. The real solution is to build on the K8s local storage feature as described in #796. In the meantime, setting the host path is a necessary tradeoff.

With the host path set, the challenge is that if the cluster is torn down and a new cluster created, the path on the host must be deleted. Otherwise, stale keys will be used to start the services and they will fail. At a minimum we can have an error message logged so it is clear why the daemon is not starting. We could potentially have the rook agent clean up this directory, but hold off on that change for now since we will be moving to local storage soon.

### PR DESCRIPTION
**Pull request by travisn**
To survive pod and node restarts, the config dir needs to be persisted to the host. As an interim change until we get the full solution on top of the local storage (#796), this change will by default set the `dataDirHostPath` to `/var/lib/rook` so the default configuration will allow for restarts. If `/var/lib/rook` is not valid for the host, the mons will fail to start and the cluster will not initialize. 

If a test cluster is destroyed and a new one brought up in its place, the path on the host must be deleted. Otherwise, the new cluster will fail to start. The mon log will have an error such as the following and enter a crash backoff loop.
```
new mon keyring does not match existing keyring in /var/lib/rook/rook-ceph-mon0/data/keyring
```

Fixes #1155

### CODE REVIEW

File: Documentation/cluster-crd.md [hunk 1]
```diff
@@ -21,7 +21,9 @@ Settings can be specified at the global level to apply to the cluster as a whole
...
-- `dataDirHostPath`: The host path where config and data should be stored for each of the services. If the directory does not exist, it will be created. Because this directory persists on the host, it will remain after pods are deleted.  Therefore, for test scenarios, the path must be deleted if you are going to delete a cluster and start a new cluster on the same hosts.  More details can be found in the Kubernetes [host path docs](https://kubernetes.io/docs/concepts/storage/volumes/#hostpath).
+- `dataDirHostPath`: The path on the host ([hostPath](https://kubernetes.io/docs/concepts/storage/volumes/#hostpath)) where config and data should be stored for each of the services. If the directory does not exist, it will be created. Because this directory persists on the host, it will remain after pods are deleted.
+  - If a path is not specified, an [empty dir](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir) will be used and the config will be lost when the pod or host is restarted.
```
Comment (kokhang) [inline Documentation/cluster-crd.md]: I would add `Not recommended` at the end of this sentence to clarify that using emptyDir is not recommended

File: Documentation/cluster-crd.md [hunk 2]
```diff
@@ -21,7 +21,9 @@ Settings can be specified at the global level to apply to the cluster as a whole
...
+- `dataDirHostPath`: The path on the host ([hostPath](https://kubernetes.io/docs/concepts/storage/volumes/#hostpath)) where config and data should be stored for each of the services. If the directory does not exist, it will be created. Because this directory persists on the host, it will remain after pods are deleted.
+  - If a path is not specified, an [empty dir](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir) will be used and the config will be lost when the pod or host is restarted. 
+  - **WARNING**: For test scenarios, if you delete a cluster and start a new cluster on the same hosts, the path used by `dataDirHostPath` must be deleted. Otherwise, stale keys and other config will remain from the previous cluster and the new mons will fail to start.
```
Comment (kokhang) [inline Documentation/cluster-crd.md]: should this be handled during a cluster-delete? So that this directory gets deleted when a cluster deleted. Similar to what @jbw976 is doing with bound volumes.

File: cluster/examples/kubernetes/rook-cluster.yaml [hunk 3]
```diff
@@ -10,7 +10,9 @@ metadata:
   namespace: rook
 spec:
   versionTag: master
-  dataDirHostPath:
+  # the path on the host where configuration files will be persisted.
+  # Important: if you reinstall the cluster, make sure you delete this directory from each host or else the mons will fail to start on the new cluster.
```
Comment (kokhang) [inline cluster/examples/kubernetes/rook-cluster.yaml]: Can you add that is left empty, it will use emptyDir?

File: cmd/rook/mon.go [hunk 4]
```diff
@@ -88,3 +95,35 @@ func startMon(cmd *cobra.Command, args []string) error {
...
+	if _, err := os.Stat(cachedKeyringFile); os.IsNotExist(err) {
+		// the mon is starting for the first time
+		logger.Infof("mon keyring is not yet cached at %s", cachedKeyringFile)
+		return nil
+	}
+
+	contents, err := ini.Load(cachedKeyringFile)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to read config file %s. %+v. Skipping check for cached keyring.\n", cachedKeyringFile, err)
+		return nil
```
Comment (kokhang) [inline cmd/rook/mon.go]: should you return the error here? and also down below where you are getting an error.

File: tests/scripts/minikube.sh [hunk 5]
```diff
@@ -81,6 +81,8 @@ case "${1:-}" in
     minikube start --memory=3000 --kubernetes-version ${KUBE_VERSION} --extra-config=apiserver.Authorization.Mode=RBAC
     wait_for_ssh
     enable_roles_for_RBAC
+    minikube ssh "sudo mkdir /mnt/sda1/var/lib/rook;sudo ln -s /mnt/sda1/var/lib/rook /var/lib/rook"
```
Comment (kokhang) [inline tests/scripts/minikube.sh]: why is this needed?

File: Documentation/cluster-crd.md (hunk 1 above)
Comment (travisn) [reply Documentation/cluster-crd.md]: sure, i'll add that. I was thinking it was an implicit recommendation, but doesn't hurt to make it explicit.

File: Documentation/cluster-crd.md (hunk 2 above)
Comment (travisn) [reply Documentation/cluster-crd.md]: the operator can't do it. We would have to send an instruction to the agents to clean it up, but not sure it's worth it, and it's out of scope for this PR

File: cluster/examples/kubernetes/rook-cluster.yaml (hunk 3 above)
Comment (travisn) [reply cluster/examples/kubernetes/rook-cluster.yaml]: will do

File: cmd/rook/mon.go (hunk 4 above)
Comment (travisn) [reply cmd/rook/mon.go]: i wanted to be conservative and only fail in the exact condition that is being checked. If there is still an error condition, it will fail later.

File: tests/scripts/minikube.sh (hunk 5 above)
Comment (travisn) [reply tests/scripts/minikube.sh]: The default path of `/var/lib/rook` will only work for our minikube environment if we create this link

File: tests/scripts/minikube.sh [hunk 6]
```diff
@@ -81,6 +81,10 @@ case "${1:-}" in
     minikube start --memory=3000 --kubernetes-version ${KUBE_VERSION} --extra-config=apiserver.Authorization.Mode=RBAC
     wait_for_ssh
     enable_roles_for_RBAC
+
+    # create a link so the default dataDirHostPath will work for this environment
+    minikube ssh "sudo mkdir /mnt/sda1/var/lib/rook;sudo ln -s /mnt/sda1/var/lib/rook /var/lib/rook"
```
Comment (dab-q) [inline tests/scripts/minikube.sh]: Shouldn't that be `mkdir -p ...`, or will `/mnt/sda1/var/lib` always pre-exist?

Comment (travisn) [reply tests/scripts/minikube.sh]: yes, it's expected to pre-exist

**Comment by jbw976** [comment]
do we already have an issue tracking the rook-agent performing clean up of the `dataDirHostPath`?  that may be an easy and nice addition to make this more smooth.

File: Documentation/cluster-crd.md [hunk 7]
```diff
@@ -21,7 +21,9 @@ Settings can be specified at the global level to apply to the cluster as a whole
...
-- `dataDirHostPath`: The host path where config and data should be stored for each of the services. If the directory does not exist, it will be created. Because this directory persists on the host, it will remain after pods are deleted.  Therefore, for test scenarios, the path must be deleted if you are going to delete a cluster and start a new cluster on the same hosts.  More details can be found in the Kubernetes [host path docs](https://kubernetes.io/docs/concepts/storage/volumes/#hostpath).
+- `dataDirHostPath`: The path on the host ([hostPath](https://kubernetes.io/docs/concepts/storage/volumes/#hostpath)) where config and data should be stored for each of the services. If the directory does not exist, it will be created. Because this directory persists on the host, it will remain after pods are deleted.
+  - If a path is not specified, an [empty dir](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir) will be used and the config will be lost when the pod or host is restarted. This option is not recommended.
```
Comment (jbw976) [inline Documentation/cluster-crd.md]: maybe bold for **not recommended**?

File: cmd/rook/mon_test.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+	  key = %s
+	`, keyring)
+	dataPath := path.Join(configDir, "data")
+	keyringPath := path.Join(dataPath, "keyring")
+	err := os.MkdirAll(dataPath, 0755)
+	assert.Nil(t, err)
+	err = ioutil.WriteFile(keyringPath, []byte(contents), 0644)
+	assert.Nil(t, err)
+
+	err = compareMonSecret(keyring, configDir)
```
Comment (jbw976) [inline cmd/rook/mon_test.go]: add a negative test too (where the cached value doesn't match the given keyring)?

File: cmd/rook/mon.go [hunk 9]
```diff
@@ -71,6 +73,11 @@ func startMon(cmd *cobra.Command, args []string) error {
 		return fmt.Errorf("missing mon port")
 	}
 
+	if err := compareMonSecret(clusterInfo.MonitorSecret, path.Join(cfg.dataDir, monName)); err != nil {
+		fmt.Fprintln(os.Stderr, err)
```
Comment (jbw976) [inline cmd/rook/mon.go]: should we be writing this to the termination log also? https://kubernetes.io/docs/tasks/debug-application-cluster/determine-reason-pod-failure/#writing-and-reading-a-termination-message.  that way the termination message would make it to the pod status visible with `kubectl`.

Also consider adding a new section to the common problems page and providing a link in this error message to that page.

File: Documentation/cluster-crd.md (hunk 7 above)
Comment (travisn) [reply Documentation/cluster-crd.md]: will do

File: cmd/rook/mon.go (hunk 9 above)
Comment (travisn) [reply cmd/rook/mon.go]: sweet, good idea

File: cmd/rook/mon_test.go (hunk 8 above)
Comment (travisn) [reply cmd/rook/mon_test.go]: hmm, negative test was intended

**Comment by travisn** [comment]
@jbw976 all the errors are now written to the termination log, PTAL

**Comment by jbw976** [comment]
do you think it's worth adding an entry to the common problems page too?  my gut feeling is that it would be worth an entry there (as well as linking to the page/section from the error message that the monitor prints when it finds a keyring mismatch)

**Comment by travisn** [comment]
@jbw976 in my last push i added better instructions on how to deal with the issue. My opinion is that error messages coming from a daemon wouldn't link to documentation, but should have enough detail in the error that if you search online, the first hit will be the documentation page. So I think this error message will fulfill that.

**Comment by travisn** [comment]
@jbw976 docs updated in the common problems page. That topic was already there, but needed to be updated with the new error messages

File: Documentation/common-problems.md [hunk 10]
```diff
@@ -234,26 +234,46 @@ What is happening here is that the MON pods are restarting and one or more of th
...
 ```
 
 ## Solution
-This is a common problem when reinitializing the Rook cluster and the local directory used for persistence has not been purged. This directory is the same directory reference by the `dataDiskHostPath` setting in the cluster CRD and is typically set to `/var/lib/rook`. To remedy simply shutdown all components of Rook and then delete the contents of `/var/lib/rook` or the directory specified by `dataDiskHostPath` on each of the hosts in the cluster. This time when the cluster CRD is applied, the rook-operator should start all the Pods as expected.
+This is a common problem reinitializing the Rook cluster when the local directory used for persistence has **not** been purged. This directory is the `dataDiskHostPath` setting in the cluster CRD and is typically set to `/var/lib/rook`. To fix the issue you will need to delete all components of Rook and then delete the contents of `/var/lib/rook` (or the directory specified by `dataDiskHostPath`) on each of the hosts in the cluster. Then when the cluster CRD is applied to start a new cluster, the rook-operator should start all the pods as expected.
```
Comment (jbw976) [inline Documentation/common-problems.md]: `dataDirHostPath` is the correct field name.  Please update here in all places.

Comment (travisn) [reply Documentation/common-problems.md]: oh ha, i remember commenting on that in the original PR for this topic, but it fell through somehow
